lib/Dialect/FIRRTL/Transforms/LowerOpenAggs.cpp

```diff
@@ -0,0 +1,629 @@
+struct NonHWField {
+  /// Type of the field, not a hardware type.
+  FIRRTLType type;
+  /// FieldID relative to base of converted type.
+  uint64_t fieldID;
+  /// Relative orientation.  False means aligned.
+  bool isFlip;
+  /// String suffix naming this field.
+  SmallString<16> suffix;
+  /// Print this structure to llvm::errs().
+  void dump() const;
+};
+
+/// Mapped port info
+struct PortMappingInfo {
+  /// Preserve this port, map use of old directly to new.
+  bool identity;
+
+  // When not identity, the port will be split:
+
+  /// Type of the hardware-only portion.  May be null, indicating all non-hw.
+  Type hwType;
+  /// List of the individual non-hw fields to be split out.
+  SmallVector<NonHWField, 0> fields;
+
+  /// Determine number of types this argument maps to.
+  size_t count(bool includeErased = false) const {
+    if (identity)
+      return 1;
+    return fields.size() + (hwType ? 1 : 0) + (includeErased ? 1 : 0);
+  }
+  void dump() const;
+};
+
+llvm::raw_ostream &operator<<(llvm::raw_ostream &os, const NonHWField &field) {
+  return os << llvm::formatv(
+             "non-HW(type={0}, fieldID={1}, isFlip={2}, suffix={3})",
+             field.type, field.fieldID, field.isFlip, field.suffix);
+}
+
+llvm::raw_ostream &operator<<(llvm::raw_ostream &os,
+                              const PortMappingInfo &pmi) {
+  if (pmi.identity)
+    return os << "(identity)";
+
+  os << "[[hw portion: ";
+  if (pmi.hwType)
+    os << pmi.hwType;
+  else
+    os << "(none)";
+  os << ", fields: ";
+  llvm::interleaveComma(pmi.fields, os);
+  return os << "]]";
+}
+
+} // namespace
+
+void NonHWField::dump() const { llvm::errs() << *this; }
+void PortMappingInfo::dump() const { llvm::errs() << *this; }
+
+template <typename Range>
+LogicalResult walkPortMappings(
+    Range &&range, bool includeErased,
+    llvm::function_ref<LogicalResult(size_t, PortMappingInfo &, size_t)>
+        callback) {
+  size_t count = 0;
+  for (const auto &[index, pmi] : llvm::enumerate(range)) {
+    if (failed(callback(index, pmi, count)))
+      return failure();
+    count += pmi.count(includeErased);
+  }
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// Visitor
+//===----------------------------------------------------------------------===//
+
+namespace {
+class Visitor : public FIRRTLVisitor<Visitor, LogicalResult> {
+public:
+  explicit Visitor(MLIRContext *context) : context(context){};
+
+  /// Entrypoint.
+  LogicalResult visit(FModuleLike mod);
+
+  using FIRRTLVisitor<Visitor, LogicalResult>::visitDecl;
+  using FIRRTLVisitor<Visitor, LogicalResult>::visitExpr;
+  using FIRRTLVisitor<Visitor, LogicalResult>::visitStmt;
+
+  LogicalResult visitDecl(InstanceOp op);
+
+  LogicalResult visitExpr(OpenSubfieldOp op);
+  LogicalResult visitExpr(OpenSubindexOp op);
+
+  LogicalResult visitUnhandledOp(Operation *op) {
+    auto notOpenAggType = [](auto type) {
```

Comment:
This is testing for a ref type, not an open agg type, so should it be called something like `notRefContainingType`?

Reply:
Ah, okay.  Was torn between enumerating the "open agg" types here vs rejecting anything open-agg-like, defined by containing references but not being a ref type (yet being a FIRRTL type).  The name is confusing, I'll just check those types.  Thanks!